snaps_tests/__init__.py

```diff
@@ -138,15 +154,41 @@ def _set_up_qemu_testbed(self):
 
     def build_snap(self, snap_content_dir):
         working_dir = os.path.join(self.src_dir, snap_content_dir)
-        subprocess.check_call(
-            [self.snapcraft_command, 'clean'], cwd=working_dir)
+        self._clean(working_dir)
+        self._snap(working_dir)
+
+    def _clean(self, project_dir):
+        command = self.snapcraft_command + ' clean'
+        self._run_command(command, project_dir)
+
+    def _snap(self, project_dir):
+        command = self.snapcraft_command + ' snap'
+        self._run_command(
+            command, project_dir, expect='Snapped .*\.snap', timeout=1200)
+
+    def _run_command(
+            self, command, working_dir, expect=pexpect.EOF, timeout=30):
+        print(command)
+        process = pexpect.spawn(
+            command, cwd=working_dir, timeout=timeout)
+        process.logfile_read = sys.stdout.buffer
         try:
-            subprocess.check_output(
-                [self.snapcraft_command, 'snap'], cwd=working_dir,
-                stderr=subprocess.STDOUT)
-        except subprocess.CalledProcessError as e:
-            self.addDetail('output', content.text_content(str(e.output)))
-            raise
+            process.expect(expect)
+        except pexpect.ExceptionPexpect:
+            self._add_output_detail(process.before)
+            raise CommandError(
+                command, working_dir,
+                'Expected output "{}" not found.'.format(expect)) from None
+        finally:
+            process.close()
+            if process.exitstatus:
+                self._add_output_detail(process.before)
+                raise CommandError(
+                    command, working_dir,
+                    'Exit status: {}.'.format(process.exitstatus))
+
+    def _add_output_detail(self, output):
+        self.addDetail('output', content.text_content(str(output)))
```

Comment:
I think it doesn't even need the str. pexpect's before returns a string, I added it to be extra safe.